p2p/async-buffer/src/lib.rs

```diff
@@ -0,0 +1,198 @@
+        // wake the sink.
+        self.sink_waker.wake();
+
+        Poll::Ready(Some(item))
+    }
+}
+
+/// The appender/sink side of the buffer.
+pub struct BufferAppender<T> {
+    /// The internal queue of items.
+    queue: UnboundedSender<(T, usize)>,
+    /// Our waker.
+    sink_waker: Arc<AtomicWaker>,
+    /// The current capacity of the buffer.
+    capacity: Arc<AtomicUsize>,
+    /// The max weight of an item, equal to the total allowed weight of the buffer.
+    max_item_weight: usize,
+}
+
+impl<T> BufferAppender<T> {
+    /// Returns a future that resolves when the channel has enough capacity for
+    /// a single message of `size_needed`.
+    ///
+    /// It should be noted that if there are no items in the buffer then a single item of any capacity is accepted.
+    /// i.e. if the capacity is 5 and there are no items in the buffer then any item even if it's weight is >5 will be
+    /// accepted.
+    pub fn ready(&mut self, size_needed: usize) -> BufferSinkReady<'_, T> {
+        let size_needed = min(self.max_item_weight, size_needed);
+
+        BufferSinkReady {
+            sink: self,
+            size_needed,
+        }
+    }
+
+    /// Attempts to add an item to the buffer returning an error if there is not enough capacity or
+    /// the [`BufferStream`] was dropped.
+    pub fn try_send(&mut self, item: T, size_needed: usize) -> Result<(), BufferError> {
+        let size_needed = min(self.max_item_weight, size_needed);
+
+        if self.capacity.load(Ordering::Acquire) < size_needed {
+            return Err(BufferError::NotEnoughCapacity);
+        }
+
+        let prev_size = self.capacity.fetch_sub(size_needed, Ordering::AcqRel);
+
+        // make sure we haven't wrapped the capacity around.
+        assert!(prev_size >= size_needed);
+
+        self.queue
+            .unbounded_send((item, size_needed))
+            .map_err(|_| BufferError::Disconnected)?;
+
+        Ok(())
+    }
+
+    /// Waits for capacity in the buffer and then sends the item.
+    pub fn send(&mut self, item: T, size_needed: usize) -> BufferSinkSend<'_, T> {
+        BufferSinkSend {
+            ready: self.ready(size_needed),
+            item: Some(item),
+        }
+    }
+}
+
+/// A [`Future`] for adding an item to the buffer.
+#[pin_project::pin_project]
+pub struct BufferSinkSend<'a, T> {
+    /// A future that resolves when the channel has capacity.
+    #[pin]
+    ready: BufferSinkReady<'a, T>,
+    /// The item to send.
+    item: Option<T>,
+}
+
+impl<'a, T> Future for BufferSinkSend<'a, T> {
+    type Output = Result<(), BufferError>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut this = self.project();
+
+        let size_needed = this.ready.size_needed;
+
+        this.ready.as_mut().poll(cx).map(|_| {
+            this.ready
+                .sink
+                .try_send(this.item.take().unwrap(), size_needed)
+        })
```

Comment:
Can you explain in `.expect()` or a comment why `item` is `Option<T>` + `.take()`n?